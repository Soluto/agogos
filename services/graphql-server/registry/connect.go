package registry

import (
	agogos "agogos/generated"
	"context"
	"io"
	"os"
	"time"

	log "github.com/sirupsen/logrus"
	"google.golang.org/grpc"
)

var (
	address = os.Getenv("REGISTRY_URL")
)

func closeWithContext(ctx context.Context, cls io.Closer) {
	<-ctx.Done()
	err := cls.Close()
	if err != nil {
		log.WithError(err).Warningln("Error closing")
	}
}

func createSubscribeClient(ctx context.Context) (agogos.Registry_SubscribeClient, error) {
	log.Infoln("Connecting to registry")
	conn, err := grpc.DialContext(ctx, address, grpc.WithInsecure())
	if err != nil {
		log.WithError(err).Warn("Error initiating GRPC channel")
		return nil, err
	}
	log.Infoln("Registry connection initiated")

	registryClient := agogos.NewRegistryClient(conn)
	stream, err := registryClient.Subscribe(ctx, &agogos.SubscribeParams{})
	if err != nil {
		log.WithError(err).Warn("Error subscribing to registry", err)
		conn.Close()
		return nil, err
	}

	log.Infoln("Subscribed to registry")

	go closeWithContext(ctx, conn)

	return stream, nil
}

func Connect(ctx context.Context) (chan *agogos.ConfigurationMessage, chan error) {
	configCh := make(chan *agogos.ConfigurationMessage)
	errCh := make(chan error)

	go func() {
		for {
			var err error
			clientCtx, cancel := context.WithCancel(ctx)
			client, err := createSubscribeClient(clientCtx)
			if err == nil {
				for {
					var config *agogos.ConfigurationMessage
					config, err = client.Recv()

					if err == nil {
						configCh <- config
					} else {
						break
					}
				}
			}

			errCh <- err
			cancel()
			time.Sleep(10 * time.Second)
		}
	}()

	return configCh, errCh
}
