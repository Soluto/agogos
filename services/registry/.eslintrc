{
    "env": {
        "es6": true,
        "mocha": true,
        "jest/globals": true
    },
    "extends": [
        "standard"
    ],
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
        "ecmaVersion": 2018,
        "sourceType": "module"
    },
    "plugins": [
        "@typescript-eslint",
        "jest"
    ],
    "rules": {
        "@typescript-eslint/no-unused-vars": "error",
        "space-before-function-paren": "off",
        "space-in-parens": "off",
        "array-bracket-spacing": "off",
        "semi": "off",
        "@typescript-eslint/semi": ["error"],
        "jest/no-disabled-tests": "warn",
        "jest/no-focused-tests": "error",
        "jest/no-identical-title": "error",
        "jest/prefer-to-have-length": "warn",
        "jest/valid-expect": "error"
    },
    "overrides": [{
        "files": ["tests/**"],
        "rules": {
            "no-unused-expressions": "off"
        }
    }]
}