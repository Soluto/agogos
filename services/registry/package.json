{
    "name": "registry",
    "version": "1.0.0",
    "description": "Registery for the GraphQL gateway schema",
    "main": "./dist/src/index.js",
    "repository": "https://github.com/Soluto-Private/agogos/registry",
    "license": "MIT",
    "scripts": {
        "build": "tsc",
        "start": "node ./dist/src/index.js",
        "debug": "tsnd --inspect=0.0.0.0:5858 --respawn  -- ./src/index.ts",
        "test": "jest",
        "test:watch": "jest --watch",
        "lint": "eslint ./src/**/*.ts"
    },
    "dependencies": {
        "body-parser": "^1.19.0",
        "express": "^4.17.1",
        "express-pino-logger": "^4.0.0",
        "glob": "^7.1.4",
        "google-protobuf": "^3.9.0",
        "graphql": "^14.4.2",
        "graphql-tag": "^2.10.1",
        "grpc": "^1.22.2",
        "mime-types": "^2.1.24",
        "nconf": "^0.10.0",
        "node-fetch": "^2.6.0",
        "pino": "^5.13.1",
        "pino-http-print": "^1.0.1",
        "pino-pretty": "^3.2.0",
        "ramda": "^0.26.1",
        "rxjs": "^6.5.2"
    },
    "devDependencies": {
        "@types/express": "^4.17.0",
        "@types/glob": "^7.1.1",
        "@types/google-protobuf": "^3.7.1",
        "@types/jest": "^24.0.16",
        "@types/mime-types": "^2.1.0",
        "@types/nconf": "^0.10.0",
        "@types/node": "^12.6.9",
        "@types/node-fetch": "^2.5.0",
        "@types/pino": "^5.8.8",
        "@types/protobufjs": "^6.0.0",
        "@types/ramda": "^0.26.18",
        "@typescript-eslint/eslint-plugin": "2.0.0",
        "@typescript-eslint/parser": "2.0.0",
        "eslint": "^6.4.0",
        "eslint-config-standard": "14.0.1",
        "eslint-plugin-import": "2.18.2",
        "eslint-plugin-jest": "^22.17.0",
        "eslint-plugin-node": "9.1.0",
        "eslint-plugin-promise": "4.2.1",
        "eslint-plugin-standard": "4.0.1",
        "jest": "^24.9.0",
        "ts-jest": "^24.0.2",
        "ts-node-dev": "^1.0.0-pre.40",
        "typescript": "^3.5.3"
    }
}
