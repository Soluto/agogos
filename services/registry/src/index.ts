import nconf = require("nconf");
import logger, { loggingMiddleware } from "./logger";
nconf.env("__");

import * as express from "express";
import gqlSchemaRoute from "./routes/schema";
import validateRoute from "./routes/validate";
import { startGrpcServer } from "./sync/sync-grpc-server";

const PORT = process.env.PORT || 4000;

const app = express();

app.use(loggingMiddleware);

app.get("/health", (_: express.Request, res: express.Response) =>
    res.send(true),
);
app.get("/metrics", (_: express.Request, res: express.Response) =>
    res.send(true),
);

app.use("/schema", gqlSchemaRoute);
app.use("/validate", validateRoute);

app.listen({ port: PORT }, () =>
    logger.info(`ðŸš€ HTTP Server ready at http://localhost:${PORT}`),
);

startGrpcServer();
