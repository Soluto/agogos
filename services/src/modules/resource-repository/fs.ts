import {ResourceGroup, ResourceRepository} from '.';
import * as envVar from 'env-var';
import {promises as fs} from 'fs';
import * as path from 'path';
import pLimit from 'p-limit';
import * as config from '../config';
import logger from '../logger';
import {FetchLatestResult} from './types';

export class FileSystemResourceRepository implements ResourceRepository {
    protected current?: {mtime: number; rg: ResourceGroup};
    protected policyAttachmentsDirInitialized = false;
    protected policyAttachments: {[filename: string]: Buffer} = {};
    protected policyAttachmentsRefreshedAt?: Date;

    constructor(protected pathToFile: string, protected policyAttachmentsFolderPath: string) {}

    async fetchLatest(): Promise<FetchLatestResult> {
        const stats = await fs.stat(this.pathToFile);

        if (this.current && stats.mtimeMs === this.current.mtime) {
            return {isNew: false, resourceGroup: this.current.rg};
        }

        const contents = await fs.readFile(this.pathToFile, 'utf8');
        const rg = JSON.parse(contents) as ResourceGroup;
        this.current = {mtime: stats.mtimeMs, rg};

        return {isNew: true, resourceGroup: rg};
    }

    async update(rg: ResourceGroup): Promise<void> {
        await fs.writeFile(this.pathToFile, JSON.stringify(rg));
    }

    async writePolicyAttachment(filename: string, content: Buffer): Promise<void> {
        await this.initializePolicyAttachmentsDir();

        const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
        await fs.writeFile(filePath, content);
    }

    public getPolicyAttachment(filename: string): Buffer {
        return this.policyAttachments[filename];
    }

    public async initializePolicyAttachments() {
        try {
            await this.refreshPolicyAttachments();
        } catch (err) {
            logger.fatal({err}, 'Failed fetching fs policy attachments on startup');
            throw err;
        }

        setInterval(async () => {
            try {
                await this.refreshPolicyAttachments();
            } catch (err) {
                logger.error(
                    {err},
                    `Failed refreshing fs policy attachments, last successful refresh was at ${this.policyAttachmentsRefreshedAt}`
                );
            }
        }, config.resourceUpdateInterval);
    }

    private async refreshPolicyAttachments() {
        const newRefreshedAt = new Date();

        const allAttachments = await this.getPolicyAttachmentsList();
        const attachmentsToRefresh = allAttachments
            .filter(a => this.shouldRefreshPolicyAttachment(a))
            .map(a => a.filename);

        if (attachmentsToRefresh.length > 0) {
            const newAttachments = await this.getPolicyAttachments(attachmentsToRefresh);
            newAttachments.forEach(a => (this.policyAttachments[a.filename] = a.content));
        }

        this.policyAttachmentsRefreshedAt = newRefreshedAt;
    }

    private shouldRefreshPolicyAttachment({filename, updatedAt}: {filename: string; updatedAt: Date}) {
        if (!this.policyAttachments[filename]) return true;
        if (!this.policyAttachmentsRefreshedAt) return true;

        return updatedAt > this.policyAttachmentsRefreshedAt;
    }

    private async getPolicyAttachmentsList(): Promise<{filename: string; updatedAt: Date}[]> {
        const filenames = await fs.readdir(this.policyAttachmentsFolderPath);
        const limit = pLimit(10);

        return Promise.all(
            filenames.map(filename => {
                return limit(async () => {
                    const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
                    const stats = await fs.stat(filePath);

                    const updatedAt = stats.mtime;
                    return {filename, updatedAt};
                });
            })
        );
    }

    private async getPolicyAttachments(filenames: string[]): Promise<{filename: string; content: Buffer}[]> {
        const limit = pLimit(10);

        return Promise.all(
            filenames.map(filename => {
                return limit(async () => {
                    const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
                    const content = await fs.readFile(filePath);

                    return {filename, content};
                });
            })
        );
    }

    private async initializePolicyAttachmentsDir() {
        if (this.policyAttachmentsDirInitialized) return;

        await fs.mkdir(this.policyAttachmentsFolderPath, {recursive: true});
        this.policyAttachmentsDirInitialized = true;
    }

    static fromEnvironment() {
        const pathToFile = envVar
            .get('FS_RESOURCE_REPOSITORY_PATH')
            .required()
            .asString();
        const policyAttachmentsFolderPath = envVar
            .get('FS_REPOSITORY_POLICY_ATTACHMENTS_FOLDER_PATH')
            .required()
            .asString();

        return new FileSystemResourceRepository(pathToFile, policyAttachmentsFolderPath);
    }
}
