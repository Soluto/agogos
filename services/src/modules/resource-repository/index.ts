export * from './types';

export { ResourceRepository } from './resource-repository';
export { CompositeResourceRepository } from './composite';
export { S3ResourceRepository } from './s3-repository';
export { FileSystemResourceRepository } from './fs-repository';
export { pollForUpdates } from './stream';
export { applyResourceGroupUpdates, applyResourceGroupDeletions } from './actions';
export { default as getResourceRepository } from './get-resource-repository';
