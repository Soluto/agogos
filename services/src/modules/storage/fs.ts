import * as path from 'path';
import { promises as fs } from 'fs';
import pLimit from 'p-limit';
import { Storage, FileStats, listFilesItem } from '.';

export class FileSystemStorage implements Storage {
  async fileStats(filePath: string): Promise<FileStats> {
    const stats = await fs.stat(filePath);

    return { lastModified: stats.mtime };
  }

  async readFile(filePath: string): Promise<{ content: Buffer }>;
  async readFile(filePath: string, options: { [k in any]: never }): Promise<{ content: Buffer }>;
  async readFile(filePath: string, options: { asString?: true }): Promise<{ content: string }>;
  async readFile(filePath: string, options: { asString?: boolean } = {}): Promise<{ content: Buffer | string }> {
    if (options.asString) return { content: await fs.readFile(filePath, 'utf8') };

    return { content: await fs.readFile(filePath) };
  }

  async writeFile(filePath: string, content: string | Buffer): Promise<void> {
    await fs.writeFile(filePath, content);
  }

  async deleteFile(filePath: string): Promise<void> {
    await fs.unlink(filePath);
  }

  async listFiles(folderPath: string): Promise<listFilesItem[]> {
    const filenames = await fs.readdir(folderPath);
    const limit = pLimit(10);

    return Promise.all(
      filenames.map(filename => {
        return limit(async () => {
          const filePath = path.resolve(folderPath, filename);
          const { lastModified } = await this.fileStats(filePath);

          return { filePath, lastModified };
        });
      })
    );
  }

  async mkdir(folderPath: string, options?: { recursive: true }): Promise<void> {
    await fs.mkdir(folderPath, options);
  }
}
