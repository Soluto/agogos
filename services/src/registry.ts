import {ApolloServer, gql, IResolvers} from 'apollo-server-fastify';
import * as fastify from 'fastify';
import pLimit from 'p-limit';

import {fetchAll, updateAll, ResourceGroup, applyResourceGroupUpdates} from './modules/resource-repository';
import * as config from './modules/config';
import {validateResourceGroupOrThrow} from './modules/validation';
import logger from './modules/logger';
import {handleSignals, handleUncaughtErrors} from './modules/shutdownHandler';

const typeDefs = gql`
    # General
    input ResourceMetadataInput {
        namespace: String!
        name: String!
    }

    type Result {
        success: Boolean!
    }

    input ResourceGroupInput {
        schemas: [SchemaInput!]
        upstreams: [UpstreamInput!]
        upstreamClientCredentials: [UpstreamClientCredentialsInput!]
    }

    type Query {
        validateResourceGroup(input: ResourceGroupInput!): Result
        validateSchemas(input: [SchemaInput!]!): Result
        validateUpstreams(input: [UpstreamInput!]!): Result
        validateUpstreamClientCredentials(input: [UpstreamClientCredentialsInput!]!): Result
    }

    type Mutation {
        updateResourceGroup(input: ResourceGroupInput!): Result
        updateSchemas(input: [SchemaInput!]!): Result
        updateUpstreams(input: [UpstreamInput!]!): Result
        updateUpstreamClientCredentials(input: [UpstreamClientCredentialsInput!]!): Result
    }

    # Schemas
    input SchemaInput {
        metadata: ResourceMetadataInput!
        schema: String!
    }

    # Upstreams
    enum AuthType {
        ActiveDirectory
    }

    """
    GraphQL doesn't support unions for input types, otherwise this would be a union of different auth types.
    Instead, the AuthType enum indicates which auth type is needed, and there's a property which corresponds to each auth type, which we validate in the registry.
    """
    input AuthInput {
        type: AuthType!
        activeDirectory: ActiveDirectoryAuthInput!
    }

    input ActiveDirectoryAuthInput {
        authority: String!
        resource: String!
    }

    input UpstreamInput {
        metadata: ResourceMetadataInput!
        host: String!
        auth: AuthInput!
    }

    # Upstream client credentials
    input ActiveDirectoryCredentials {
        authority: String!
        clientId: String!
        clientSecret: String!
    }

    """
    GraphQL doesn't support unions for input types, otherwise this would be a union of different auth types.
    Instead, the AuthType enum indicates which auth type is needed, and there's a property which corresponds to each auth type, which we validate in the registry.
    """
    input UpstreamClientCredentialsInput {
        metadata: ResourceMetadataInput!
        authType: AuthType!
        activeDirectory: ActiveDirectoryCredentials!
    }
`;

interface ResourceMetadataInput {
    namespace: string;
    name: string;
}

interface ResourceGroupInput {
    schemas?: SchemaInput[];
    upstreams?: UpstreamInput[];
    upstreamClientCredentials?: UpstreamClientCredentialsInput[];
}

interface SchemaInput {
    metadata: ResourceMetadataInput;
    schema: string;
}

export enum AuthType {
    ActiveDirectory = 'ActiveDirectory',
}

interface ActiveDirectoryAuthInput {
    authority: string;
    resource: string;
}

interface UpstreamInput {
    metadata: ResourceMetadataInput;
    host: string;
    auth: {
        type: AuthType;
        activeDirectory: ActiveDirectoryAuthInput;
    };
}

interface UpstreamClientCredentialsInput {
    metadata: ResourceMetadataInput;
    authType: AuthType;
    activeDirectory: {
        authority: string;
        clientId: string;
        clientSecret: string;
    };
}

async function fetchAndValidate(updates: Partial<ResourceGroup>): Promise<ResourceGroup> {
    const rg = await fetchAll();
    const newRg = applyResourceGroupUpdates(rg, updates);
    validateResourceGroupOrThrow(newRg);

    return newRg;
}

const singleton = pLimit(1);
const resolvers: IResolvers = {
    Query: {
        async validateResourceGroup(_, args: {input: ResourceGroupInput}) {
            await fetchAndValidate(args.input);

            return {success: true};
        },
        async validateSchemas(_, args: {input: SchemaInput[]}) {
            await fetchAndValidate({schemas: args.input});

            return {success: true};
        },
        async validateUpstreams(_, args: {input: UpstreamInput[]}) {
            await fetchAndValidate({upstreams: args.input});

            return {success: true};
        },
        async validateUpstreamClientCredentials(_, args: {input: UpstreamClientCredentialsInput[]}) {
            await fetchAndValidate({upstreamClientCredentials: args.input});

            return {success: true};
        },
    },
    Mutation: {
        updateResourceGroup(_, args: {input: ResourceGroupInput}) {
            return singleton(async () => {
                const rg = await fetchAndValidate(args.input);
                await updateAll(rg);

                return {success: true};
            });
        },
        updateSchemas(_, args: {input: SchemaInput[]}) {
            return singleton(async () => {
                const rg = await fetchAndValidate({schemas: args.input});
                await updateAll(rg);

                return {success: true};
            });
        },
        updateUpstreams(_, args: {input: UpstreamInput[]}) {
            return singleton(async () => {
                const rg = await fetchAndValidate({upstreams: args.input});
                await updateAll(rg);

                return {success: true};
            });
        },
        updateUpstreamClientCredentials(_, args: {input: UpstreamClientCredentialsInput[]}) {
            return singleton(async () => {
                const rg = await fetchAndValidate({upstreamClientCredentials: args.input});
                await updateAll(rg);

                return {success: true};
            });
        },
    },
};

export const app = new ApolloServer({
    typeDefs,
    resolvers,
    tracing: config.enableGraphQLTracing,
    playground: config.enableGraphQLPlayground,
    introspection: config.enableGraphQLIntrospection,
});

async function run() {
    logger.info('Stitch registry booting up...');
    const server = fastify();
    server.register(app.createHandler({path: '/graphql'}));
    const address = await server.listen(config.httpPort, '0.0.0.0');
    logger.info({address}, 'Stitch registry started');

    handleSignals(() => app.stop());
    handleUncaughtErrors();
}

// Only run when file is being executed, not when being imported
if (require.main === module) {
    run();
}
