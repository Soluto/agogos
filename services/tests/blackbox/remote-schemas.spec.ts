import * as fs from 'fs/promises';
import * as nock from 'nock';
import { FastifyInstance } from 'fastify';
import { graphqlSync, print } from 'graphql';
import { makeExecutableSchema } from 'graphql-tools';
import { gql, GraphQLRequest } from 'apollo-server-core';
import { createServer as createRegistry } from '../../src/registry';
import GraphQLErrorSerializer from '../utils/graphql-error-serializer';
import { ResourceGroup, ResourceMetadata, Schema } from '../../src/modules/resource-repository';

describe('Remote Schemas', () => {
  const remoteServer = 'http://remote-server';

  let app: FastifyInstance;
  let dispose: () => Promise<void>;

  const metadata: ResourceMetadata = { namespace: 'remote-schemas', name: 'schema' };

  const schema: Schema = {
    metadata,
    schema: print(gql`
      type Query {
        foo: String! @gql(url: "${remoteServer}/graphql", fieldName: "foo")
      }
  `),
  };

  const remoteSchemasPayload: GraphQLRequest = {
    query: print(gql`
      query {
        remoteSchemas {
          url
          schema
        }
      }
    `),
  };

  const remoteSchemaPayload: GraphQLRequest = {
    query: print(gql`
      query($url: String!) {
        remoteSchema(url: $url) {
          url
          schema
        }
      }
    `),
    variables: {
      url: `${remoteServer}/graphql`,
    },
  };

  beforeAll(async () => {
    expect.addSnapshotSerializer(GraphQLErrorSerializer);

    const remoteSchema = makeExecutableSchema({
      typeDefs: gql`
        type Query {
          foo: String
        }
      `,
      resolvers: {
        Query: {
          foo: () => 'FOO',
        },
      },
    });

    nock(remoteServer)
      .post('/graphql')
      .reply(200, (_, body: any) =>
        graphqlSync({
          schema: remoteSchema,
          source: body.query,
          variableValues: body.variables,
          operationName: body.operationName,
        })
      );

    const resources: ResourceGroup = {
      schemas: [],
      upstreams: [],
      upstreamClientCredentials: [],
      policies: [],
      remoteSchemas: [],
    };

    await fs.writeFile(process.env.FS_RESOURCE_REPOSITORY_PATH!, JSON.stringify(resources));
    await fs.writeFile(process.env.FS_REGISTRY_RESOURCE_REPOSITORY_PATH!, JSON.stringify(resources));

    app = await createRegistry();
  });

  afterAll(async () => {
    await dispose();
    await fs.unlink(process.env.FS_RESOURCE_REPOSITORY_PATH!);
    await fs.unlink(process.env.FS_REGISTRY_RESOURCE_REPOSITORY_PATH!);
  });

  test('Query empty resource group', async () => {
    const response1 = await app.inject({
      method: 'POST',
      url: '/graphql',
      payload: remoteSchemasPayload,
    });

    expect(response1.statusCode).toEqual(200);
    expect(response1.json().data).toMatchSnapshot('remote schemas');

    const response2 = await app.inject({
      method: 'POST',
      url: '/graphql',
      payload: remoteSchemaPayload,
    });

    expect(response2.statusCode).toEqual(200);
    expect(response2.json().data).toMatchSnapshot('remote schema');
  });

  test('Add schema with @gql directive', async () => {
    const payload: GraphQLRequest = {
      query: print(gql`
        mutation UpdateSchemas($input: [SchemaInput!]!) {
          updateSchemas(input: $input) {
            success
          }
        }
      `),
      variables: {
        input: [schema],
      },
    };

    const response = await app.inject({
      method: 'POST',
      url: '/graphql',
      payload,
    });

    expect(response.statusCode).toEqual(200);
    expect(response.json().data).toMatchSnapshot();
  });

  test('Query resource group', async () => {
    const response1 = await app.inject({
      method: 'POST',
      url: '/graphql',
      payload: remoteSchemasPayload,
    });

    expect(response1.statusCode).toEqual(200);
    expect(response1.json().data).toMatchSnapshot('remote schemas');

    const response2 = await app.inject({
      method: 'POST',
      url: '/graphql',
      payload: remoteSchemaPayload,
    });

    expect(response2.statusCode).toEqual(200);
    expect(response2.json().data).toMatchSnapshot('remote schema');
  });

  test('Delete the schema with @gql directive', async () => {
    const payload: GraphQLRequest = {
      query: print(gql`
        mutation DeleteSchemas($input: [ResourceMetadataInput!]!) {
          deleteSchemas(input: $input) {
            success
          }
        }
      `),
      variables: {
        input: [metadata],
      },
    };

    const response = await app.inject({
      method: 'POST',
      url: '/graphql',
      payload,
    });

    expect(response.statusCode).toEqual(200);
    expect(response.json().data).toMatchSnapshot();
  });

  test('Query resource group after deletion', async () => {
    const response1 = await app.inject({
      method: 'POST',
      url: '/graphql',
      payload: remoteSchemasPayload,
    });

    expect(response1.statusCode).toEqual(200);
    expect(response1.json().data).toMatchSnapshot('remote schemas');

    const response2 = await app.inject({
      method: 'POST',
      url: '/graphql',
      payload: remoteSchemaPayload,
    });

    expect(response2.statusCode).toEqual(200);
    expect(response2.json().data).toMatchSnapshot('remote schema');
  });
});
