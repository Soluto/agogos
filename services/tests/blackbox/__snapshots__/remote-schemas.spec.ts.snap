// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remote Schemas Add schema with @gql directive 1`] = `
Object {
  "updateSchemas": Object {
    "success": true,
  },
}
`;

exports[`Remote Schemas Add schema with @gql directive: gateway resource group 1`] = `
Array [
  Object {
    "schema": "type Query {
  foo: String
}
",
    "url": "http://remote-server/graphql",
  },
]
`;

exports[`Remote Schemas Add schema with @gql directive: registry resource group 1`] = `
Array [
  Object {
    "schema": "type Query {
  foo: String
}
",
    "url": "http://remote-server/graphql",
  },
]
`;

exports[`Remote Schemas Delete the schema with @gql directive 1`] = `
Object {
  "deleteSchemas": Object {
    "success": true,
  },
}
`;

exports[`Remote Schemas Query empty resource group: remote schema 1`] = `
Object {
  "remoteSchema": null,
}
`;

exports[`Remote Schemas Query empty resource group: remote schemas 1`] = `
Object {
  "remoteSchemas": Array [],
}
`;

exports[`Remote Schemas Query resource group after deletion: remote schema 1`] = `
Object {
  "remoteSchema": Object {
    "schema": "type Query {
  foo: Int
}
",
    "url": "http://remote-server/graphql",
  },
}
`;

exports[`Remote Schemas Query resource group after deletion: remote schemas 1`] = `
Object {
  "remoteSchemas": Array [
    Object {
      "schema": "type Query {
  foo: Int
}
",
      "url": "http://remote-server/graphql",
    },
  ],
}
`;

exports[`Remote Schemas Query resource group again: remote schema 1`] = `
Object {
  "remoteSchema": Object {
    "schema": "type Query {
  foo: Int
}
",
    "url": "http://remote-server/graphql",
  },
}
`;

exports[`Remote Schemas Query resource group: remote schema 1`] = `
Object {
  "remoteSchema": Object {
    "schema": "type Query {
  foo: String
}
",
    "url": "http://remote-server/graphql",
  },
}
`;

exports[`Remote Schemas Query resource group: remote schemas 1`] = `
Object {
  "remoteSchemas": Array [
    Object {
      "schema": "type Query {
  foo: String
}
",
      "url": "http://remote-server/graphql",
    },
  ],
}
`;

exports[`Remote Schemas Refresh remote schema 1`] = `
Object {
  "refreshRemoteSchema": Object {
    "success": true,
  },
}
`;

exports[`Remote Schemas Refresh remote schema: gateway resource group 1`] = `
Array [
  Object {
    "schema": "type Query {
  foo: Int
}
",
    "url": "http://remote-server/graphql",
  },
]
`;

exports[`Remote Schemas Refresh remote schema: registry resource group 1`] = `
Array [
  Object {
    "schema": "type Query {
  foo: Int
}
",
    "url": "http://remote-server/graphql",
  },
]
`;

exports[`Remote Schemas Try to refresh remote schema for non-exiting url 1`] = `
{
  "errors": [
    {
      "message": "Remote schema refresh request failed: Error: Unknown url: http://does-not-exist",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {
    "refreshRemoteSchema": null
  }
}
`;

exports[`Remote Schemas Try to refresh remote schema when there is no remote schemas at all 1`] = `
{
  "errors": [
    {
      "message": "Remote schema refresh request failed: Error: Unknown url: http://remote-server/graphql",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {
    "refreshRemoteSchema": null
  }
}
`;
